sycl/source/detail/queue_impl.cpp

```diff
@@ -234,21 +242,42 @@ void queue_impl::wait(const detail::code_location &CodeLoc) {
   TelemetryEvent = instrumentationProlog(CodeLoc, Name, StreamID, IId);
 #endif
 
-  std::vector<std::weak_ptr<event_impl>> Events;
-  std::vector<event> USMEvents;
+  std::vector<std::weak_ptr<event_impl>> WeakEvents;
+  std::vector<event> SharedEvents;
   {
-    std::lock_guard<std::mutex> Lock(MMutex);
-    Events.swap(MEventsWeak);
-    USMEvents.swap(MEventsShared);
+    std::lock_guard<mutex_class> Lock(MMutex);
+    WeakEvents.swap(MEventsWeak);
+    SharedEvents.swap(MEventsShared);
+  }
+  // If the queue is either a host one or does not support OOO (and we use
+  // multiple in-order queues as a result of that), wait for each event
+  // directly. Otherwise, only wait for unenqueued or host task events, starting
+  // from the latest submitted task in order to minimize total amount of calls,
+  // then handle the rest with piQueueFinish.
+  for (auto EventImplWeakPtrIt = WeakEvents.rbegin();
+       EventImplWeakPtrIt != WeakEvents.rend(); ++EventImplWeakPtrIt) {
+    if (std::shared_ptr<event_impl> EventImplSharedPtr =
+            EventImplWeakPtrIt->lock()) {
+      // A nullptr PI event indicates that piQueueFinish will not cover it,
+      // either because it's a host task event or an unenqueued one.
+      if (is_host() || !MSupportOOO ||
+          nullptr == EventImplSharedPtr->getHandleRef()) {
+        EventImplSharedPtr->wait(EventImplSharedPtr);
+      }
+    }
+  }
+  if (!is_host() && MSupportOOO) {
+    const detail::plugin &Plugin = getPlugin();
+    Plugin.call<detail::PiApiKind::piQueueFinish>(getHandleRef());
+    for (std::weak_ptr<event_impl> &EventImplWeakPtr : WeakEvents)
+      if (std::shared_ptr<event_impl> EventImplSharedPtr =
+              EventImplWeakPtr.lock())
+        EventImplSharedPtr->cleanupCommand(EventImplSharedPtr);
+    assert(SharedEvents.empty());
```

Comment:
Could you, please, clarify how does `SharedEvents` vector become empty?

Reply:
The assertion is that these queues (those that are non-host and do not emulate OOO with multiple in-order queues) do not store anything in SharedEvents in the first place. I'll add an assertion message to make that intent clearer (and probably add an assertion to addSharedEvent as well).